doc/developer/design/20231213_compute_operator_frontier_tracking.md

```diff
@@ -0,0 +1,187 @@
+The reporting interval should be configurable through an LD flag, which will allow us to find a good default during testing.
+
+### SQL Relations
+
+As described in the previous section, this design proposes to add a `replica_id` column to four existing introspection relations, and to convert them from per-replica relations to regular ones.
+
+Additionally, we propose adding a new relation to expose operator frontier information:
+
+```
+mz_internal.mz_dataflow_operator_frontiers_per_worker {
+    id: uint8,
+    replica_id: text,
+    worker_id: uint8,
+    time: mz_timestamp,
+}
+```
+
+## Minimal Viable Prototype
+
+TODO
+
+## Alternatives
+
+### Progress Tracking Using Fine-grained Timestamps
+
+Another approach to measuring progress through the initial snapshot processing is to make dataflow inputs provide the snapshot data at multiple distinct timestamps.
+For example, an input could, rather than inserting 1GB of data at time `T`, insert 100MB each at times `(T, 0)`, `(T, 1)`, ..., `(T, 9)`.
+These fine-grained input timestamps would be reflected at the dataflow outputs as well and allow tracking progress at finer granularity than regular time advancements.
+
+We don't choose this approach for these reasons:
+
+* Implementation effort: Changing the main timestamp type used in Compute and accounting for all the new edge cases introduced by this change seems like a lot of effort.
+  That said, no prototype implementation was attempted, so we cannot say for sure.
+* For the above approach to produce useful progress indicators (10%, 20%, ..., 100%), it needs to be the case that each chunk of the snapshot data takes roughly the same time to process.
+  There is no reason to believe that this is true.
+  In fact, it is likely that later chunks will need longer to process in dataflows that contain joins.
+  As a result, we would probably end up reporting progress in a manner that is very confusing (imagine a progress bar that is stuck at 90% forever).
+* This approach of reporting progress does not help in finding bottlenecks in dataflow plans.
+
+## Open questions
+
+### How to expose operator hydration status?
+
+Knowing the operator frontiers is not quite enough to make statements like "x out of N operators have hydrated".
+We also need to know the per-replica as-of frontier of the dataflow, to know when frontiers have advanced past it.
+The compute controller knows all as-ofs and can make these comparisons, so the only question is how the operator hydration status should be exposed.
+
+Some options:
+
+* Add a new relation `mz_dataflow_operator_hydration_statuses`, akin to `mz_{compute_}hydration_statuses`.
+* Add a `hydrated` flag to some existing per-operator relation, like `mz_dataflow_operator_frontiers_per_worker`.
+* Expose the per-replica `as_of` value directly in some relation and then do the frontier comparison in a view.
```

Comment:
I think for what we want to expose in console, a hydrated flag is all we need. That said, I don't have a strong opinion here, but it seems like exposing the as_of and putting a view on top gives us the most flexibility?

Reply:
Indeed. I believe we could replace `mz_compute_hydration_statuses` with a view if we had the `as_of` in a table. In fact, we could even delete `mz_compute_hydration_statuses` and inline the view logic into `mz_hydration_statuses`.

Thinking about it, it should be possible to add the `as_of` to `mz_cluster_replica_frontiers`. It fits very well there.